rte/kernels/mo_rte_solver_kernels.F90

```diff
@@ -947,5 +949,295 @@ pure subroutine apply_BC_0(ncol, nlay, ngpt, top_at_1, flux_dn) bind (C, name="a
+          scaleTau = (1._wp - ssal + wb )
+          ! Eq.15 of the paper
+          tauLoc(icol, ilay, igpt) = scaleTau * tau(icol, ilay, igpt)
+          ! 
+          ! here ssa is used to store parameter wb/[1-w(1-b)] of Eq.21 of the Tang's paper
+          ! actually it is in line of parameter rescaling defined in Eq.7
+          if (scaleTau > epsilon(1._wp)) then
+            scaling(icol, ilay, igpt) = wb / scaleTau
+          else
+            scaling(icol, ilay, igpt) = 1._wp
+          endif
+        enddo
+      enddo
+    enddo
+    !$acc exit data copyout(tauLoc, scaling)
+    !$acc exit data delete(tau, ssa, g)
+  end subroutine scaling_1rescl
+
+! -------------------------------------------------------------------------------------------------
+!
+! Longwave no-scattering + a single itergation transport
+!
+! 
+! Implemented based on the paper
+! Tang G, P Yang, GW Kattawar, X Huang, EJ Mlawer, BA Baum, MD King, 2018: Improvement of 
+! the Simulation of Cloud Longwave Scattering in Broadband Radiative Transfer Models, 
+! Journal of the Atmospheric Sciences 75 (7), 2217-2233
+! https://doi.org/10.1175/JAS-D-18-0014.1
+! adustment terms is computed based on solution of the Tang equations
+! for "linear-in-tau" internal source
+! 
+! -------------------------------------------------------------------------------------------------
+subroutine lw_transport_1rescl(ncol, nlay, top_at_1, &
+                               trans, sfc_albedo, source_dn, source_up, source_sfc, &
+                               radn_up, radn_dn, An, Cn) bind(C, name="lw_transport_1rescl")
+  integer,                          intent(in   ) :: ncol, nlay ! Number of columns, layers, g-points
+  logical(wl),                      intent(in   ) :: top_at_1   !
+  real(wp), dimension(ncol,nlay  ), intent(in   ) :: trans      ! transmissivity = exp(-tau)
+  real(wp), dimension(ncol       ), intent(in   ) :: sfc_albedo ! Surface albedo
+  real(wp), dimension(ncol,nlay  ), intent(in   ) :: source_dn, &
+                                                     source_up  ! Diffuse radiation emitted by the layer
+  real(wp), dimension(ncol       ), intent(in   ) :: source_sfc ! Surface source function [W/m2]
+  real(wp), dimension(ncol,nlay+1), intent(inout) :: radn_up    ! Radiances [W/m2-str]
+  real(wp), dimension(ncol,nlay+1), intent(inout) :: radn_dn    !Top level must contain incident flux boundary condition
+  real(wp), dimension(ncol,nlay),   intent(in   ) :: An, Cn
+  ! Local variables
+  integer :: ilev, icol
+  ! ---------------------------------------------------
+  ! real(wp), dimension(ncol,nlay+1)                :: i_up    ! Radiances [W/m2-str]
+  real(wp) :: adjustmentFactor
+  if(top_at_1) then
+    !
+    ! Top of domain is index 1
+    !
+    ! 1st Upward propagation
+    do ilev = nlay, 1, -1
+      radn_up(:,ilev) = trans(:,ilev  )*radn_up(:,ilev+1) + source_up(:,ilev)
+      do icol=1,ncol
+         if ( Cn(icol,ilev) > 1e-6 )  then
```

Comment:
I've asked about these checks several times before. I'm not convinced they will save time - in a vectorized code the compiler will execute both branches and then mask out the one that isn't needed. I would greatly prefer that we remove all the `if` statements currently in place to save time. You are welcome to leave a comment in place in case we find a big performance it.

Reply:
That particular loop may be non-serializable since the result of the current iteration depends on the previous one. I removed the "if's"

Reply:
In scaling_1rescl there also an "if" statement for improbable situation when g=w=1 which cannot ne in LW region. I can remove it with comment. What do you think?

Reply:
I would prefer to have the input validated before passing it in - so if there will be an error if g = w = 1 when doing the rescaling, let's check for that, probably before even calling the solvers.

Reply:
That is a good idea, but it would only identify the potential issue not to solve it.
Currently, for validation I rely on validate_2stream that checks on 
that  0 <=ssa<=1 and -1<=g<=1 which isn't informative in this case.
Should an additional validation function added to the code?

Reply:
And actually it not an error it is just an indeterminate form 
w(1-g)/(2-w(1+g))->0/0 ->1 
if w->1 and g->1

Reply:
In DISORT or like methods similar problem is solved by allowing 0< =ssa <= 1-epsilon
where epsilon is about 1e-4 or less.

Reply:
I'm not sure what the right solution is here. I don't like the `if` statements, but neither do I want code that could crash with a division by zero, even if that only happens when the user has provided bogus input. We could have `validate()` ensure that `ssa < 1._wp-epsilon(1.wp))` but I'm not sure that's a general solution (and anyway I'd want to do the analysis to see what the rounding error is.)

Reply:
Actually we have to check only on a single condition that ssa*g < 1. If g=1 - epsilon it means that instead of Tang approximation a correctly rescaled no scattering solution would be more accurate. Let me think how it can be implemented with less pain. Your thoughts would be appreciated. Trivial solution. 
1. create two versions  of the scaling_1rescl
   a)  fast but dangerous
   b)  slow but accurate
2. Check on condition 1-epsilon < ssa*g <=1
   a) if True then call slow
   b) if False call fast

Final thought. I have to verify what solution is accurate if g->1. I suspect that Tang may be no a leader